### DISCUSSION TITLE
Separate directories for output components

### DISCUSSION THREAD

**Discussion by k18ldp** [RogueOneEcho/caesura#77] [Ideas]
As per red_oxide, allow user to configure individual directories for transcodes, spectograms and .torrent files.

In the current configuration all above components will be stored in `output` folder.

This is very useful for organisation, easy of use and decluttering.

**Comment by RogueOneEcho** [comment]
**Can you expand on why this would be useful?**

I did play around with many different directory formats for `output`, including various nesting structures but couldn't see any actual advantage to them and they just added unnecessary complexity.
1. `output/${SOURCE_NAME}/spectrograms`, `output/${SOURCE_NAME}/transcodes`, `output/${SOURCE_NAME}/torrents`
2. `output/spectrograms/${SOURCE_NAME}`, `output/transcodes/${SOURCE_NAME}`, `output/torrents/${SOURCE_NAME}`

---

Personally, I found the naming of the separate directories in `red_oxide` to be too vague `content-directory` `transcode-directory`, `torrent-directory`, `spectrogram-directory`. Particularly `torrent-directory` wasn't clear whether that was where the content of the torrent lived or the source torrent files, or where torrent files are created.

Having so many options that needed to be understood and set before first using the app wasn't a great onboarding experience.

Hence unifying to a single `output` directory makes the purpose explicit. You should still be able to declutter easily. 
- Most GUI applications will allow you to sort or group by file type so you can easily select all the `.torrent` files for example.
- From the CLI you can easily `rm output/*.torrent` or `cp output/*.torrent /path/to/autoadd/`
- Spectrograms are a little trickier, but CLI wildcards still work: `ls -1 output/*SPECTROGRAMS]`

---

From a developer perspective:
- there's also some safety that everything in the `ouput` directory has been generated by the application so there are only minimal checks whether files are being overwritten (#34).
- And users are required to copy torrent files or transcodes elsewhere to inject into clients. Therefore there's issue if something such as #76 is implemented. Whereas if users could point the torrent output directory anywhere (such as `deluge/state` or `qBittorrent/BT_backup` it may cause issues if the app deletes that file

---

Potentially a `copy_spectrogram_to` command could be useful. With that you'd then have options to copy all outputs to different locations and can confidently just routinely clear the `output` directory?

---

That was all a long rambling way of saying I don't have plans to do this at present, but I could be swayed if there was demand for it, or if there's a use case I've missed.

**Comment by k18ldp** [comment]
Hey thanks for the reply.

I can explain my use-case hopefully it will make sense, if not give me shout.

My current setup with red_oxide is for transcodes to go in a directory I have specifically for transcoded content:
eg .../torrents/music/transcodes

I also have separate directories for my _uploads_ to RED/OPS and music I'm cross-seeding:
.../torrents/music/uploads
.../torrents/music/seeding

My additional directories for `.torrent files` and `specs` go inside of the transcodes directory (as they are related to the transcoded files)
.../torrents/music/transcodes/1_specs
.../torrents/music/transcodes/2_torrents

(using the 1 and 2 just so that they are always at the top of my unraid directory gui)

Now when i processing any transcodes - I have the three folders open in finder (mac) or file explorer:
transcodes, specs, torrents.
This allows me then to quickly check (using `sort by - date added`):
1. spectograms before proceeding to transcoding
2. files that have been transcoded for any errors
3. copy the .torrent files directly into qbit when I'm ready

After processing - I'm left with all spectograms neatly in a separate `specs` directory, torrent files in a `torrents` directory and the transcoded media in the parent folder. 

 I've not done _many_ uploads/transcode currently but I want to have a good system to ensure the folders don't get out of control and are relatively easy to manage :)



Anyway that's my use case... If it's not a priority for you - that's totally understandable - I'll try and amend/adapt my workflow a little and see how I can make it work for me

**Reply by RogueOneEcho** [reply]
> Now when i processing any transcodes - I have the three folders open in finder (mac) or file explorer: transcodes, specs, torrents. This allows me then to quickly check (using `sort by - date added`):

Three folders open simultaneously definitely sounds excessive 😆. 

For my own workflow I use the `batch --transcode --spectrogram --no-limit` command to generate spectrograms and transcode everything at once. 

Then at my leisure I can check everything before upload, I find it's an unnecessary delay to check specs before transcoding.

I have the output folder mounted to a Caddy container that serves all the files. I can easily sort by date and check everything together. Spectrograms and transcodes are ordered next to one another both alphabetically and by date. And handily the browser marks the visited links purple so it's easy to keep track.

Then slowly upload them with `batch --upload --limit 5`. The `.torrent` file gets copied to an autoadd folder for my torrent client.

What I haven't really considered is the number of spectrograms I have lying around with no real use anymore. Potentially #74 would address that

**Comment by khlaus** [comment]
For me it would help to have an copy_torrent_to option for the batch command, so I don't have to go into the folder where there are 300 albums to look for the torrent, I'd prefer it to go into a /torrents or into the watch folder directly. I know that there is one for the upload command.

**Reply by RogueOneEcho** [reply]
Can you clarify what you mean? I'm pretty sure the `copy_torrent_to` option should work with the batch command. You are right however that the docs don't list it so I need to investigate https://github.com/RogueOneEcho/caesura/blob/main/COMMANDS.md#caesura-batch

**Reply by khlaus** [reply]
![image](https://ptpimg.me/217vi5.png)

**Reply by RogueOneEcho** [reply]
Oh you know what, I bet it works if you define it in the config.yml but not via the CLI. I'll look in to it soon.

**Reply by khlaus** [reply]
Yes, defining it in config.yml worked indeed!

**Reply by RogueOneEcho** [reply]
fixed in 7860d2a0 #102

**Comment by sage2050** [comment]
seconded on simply letting the user specify output directories if desired. 

Also the current workflow [(discussed here)](https://github.com/RogueOneEcho/caesura/issues/85) leaves a lot of trash around that's kind of a pain to clean up. If i'm copying the torrents to another location for auto-add why is it leaving the original behind? why is the .torrent used to check for transcodes not deleted?

**Reply by RogueOneEcho** [reply]
I've created a separate discussion for this: https://github.com/RogueOneEcho/caesura/discussions/116

I wasn't in favor of separate output directories but you raise a solid point that if the transcode is uploaded and the content copied elsewhere then it does make sense to clear the `output` directory.
